applications/mne_x/plugins/averaging/averaging.cpp

```diff
@@ -290,15 +227,158 @@ QString Averaging::getName() const
 
 //*************************************************************************************************************
 
+QWidget* Averaging::setupWidget()
+{
+    AveragingSetupWidget* setupWidget = new AveragingSetupWidget(this);//widget is later distroyed by CentralWidget - so it has to be created everytime new
+    return setupWidget;
+}
+
+
+//*************************************************************************************************************
+
+void Averaging::update(XMEASLIB::NewMeasurement::SPtr pMeasurement)
+{
+    QSharedPointer<NewRealTimeMultiSampleArray> pRTMSA = pMeasurement.dynamicCast<NewRealTimeMultiSampleArray>();
+
+    if(pRTMSA) {
+        //Check if buffer initialized
+        if(!m_pAveragingBuffer)
+            m_pAveragingBuffer = CircularMatrixBuffer<double>::SPtr(new CircularMatrixBuffer<double>(64, pRTMSA->getNumChannels(), pRTMSA->getMultiSampleArray()[0].cols()));
+
+        //Fiff information
+        if(!m_pFiffInfo)
+        {
+            m_pFiffInfo = pRTMSA->info();
+            emit fiffInfoAvailable();
+
+#ifdef DEBUG_AVERAGING
+            for(qint32 i = 0; i < m_pFiffInfo->nchan; ++i)
+            {
+                if(m_pFiffInfo->chs[i].kind == FIFFV_STIM_CH)
+                {
+                    m_iTestStimCh = i;
+                    break;
+                }
+            }
+#endif
+        }
+
+
+        if(m_bProcessData)
+        {
+            for(qint32 i = 0; i < pRTMSA->getMultiSampleArray().size(); ++i)
+            {
+                MatrixXd t_mat = pRTMSA->getMultiSampleArray()[i];
```

Comment:
It is maybe possible to use const MatrixXd& t_mat = pRT...
I am not sure about the behaviour of the semaphore, since it willl not be released. It might lead to a blocking unti the copy is perormed.